core/profile.c

```diff
@@ -1107,6 +1109,20 @@ static void calculate_deco_information(struct deco_state *ds, const struct deco_
 					entry->surface_gf = surface_gf;
 			}
 
+			// In the planner, if the ceiling is violated, add an event.
+			// TODO: This *really* shouldn't be done here. This is a contract
```

Comment:
Maybe a (historical) comment is in order: I view the ceilings only as a visualisation of the tissue loadings (which is the underlying data), so computing them in the rendering could be justified. 

One would think that the planner should not produce plans that violate the ceilings (pretty much my definition of what the planner is supposed to do) but it turns out it can: This happens mostly when using VPM-B where the deco model of the profile slightly differs from the deco model that the planner uses as the planner has more information available: in particular this the distinction between bottom phase of the dive and ascent phase which is well defined only in the planner and not for real dives. But we decided, the profile should use only information that is available also for logged dives and so it has to make a guess about that distinction (we use the point of the deepest ceiling). So it can happen, that with the information of the planner a point is below the ceiling but for the profile it is above. To understand that effect better, we wanted a visual representation of when this happens so we added the event.

Another possibility for this to happen is that for some reason (e.g. excessive computation time) the planner gives up and then one also wants to know up to which point to trust the plan. 

And yes, one could move the whole deco calculations of the profile to an earlier point (where one could still legitimately modify the dive) but I think this would end up being pretty ugly as well.

Reply:
> Maybe a (historical) comment is in order: I view the ceilings only as a visualisation of the tissue loadings (which is the underlying data), so computing them in the rendering could be justified.

I understand that. The comment was concerning the addition of an event to the dive. Instead of adding an event to the dive, the code should add a "warning at point X" flag to the `profile_info` structure.